### ISSUE TITLE
Dynamic Changelog

### ISSUE THREAD

**Issue by tlovett** [useblocks/sphinxcontrib-needs#56]
If someone is using a needs file for your prior version, it would be nice to be able to generate a summary of changes since the last version.

For example, a page that lists out all the needs that were 
- Added
- Deleted 
- Changed

It would also be useful if these changes were tracked as dynamic fields so they could be used in filterable directives.  For example, here's a table of new requirements.

```rst
.. needtable::
   :columns: id,title,change_type
   :filter: 'added' in change_type
```
Need to think of how this would be formatted, but it might be a new directive.  

We're actively looking at creating this on every release so we're interested in automating this.

**Comment by danwos** [comment]
I'm not really sure how high the minimum effort will be to get this done or if we even should aim higher, to make sphinx-needs capable to handle historical data, which gets imported from multiple needs.json files.

Currently sphinx-needs knows nothing about old or new versions of a need. 
There is a single flat list, which contains all valid needs, which are used in the current rst docs.
Being able to get the differences of need versions means to save need-data, which is not used directly in the rst.

I see a lot of use cases, which would be realizable, if sphinx-needs can handle and provide needs based on their version (added/deleted, diff of content, highlight if status has changed, selection field on a need to dynamically load a different version, ...).

But all this means three changes/improvements for me:
**Add an additional layer to all needs**
The need-list must look something like
```python
needlist = {
    "NEED_ID_A": {
         "CURRENT": needinfo,
         "VERSION_1": needinfo,
         "VERSION_2": needinfo,
   },
   # Was not part of a specific version in the past
    "NEED_ID_B": {
         "CURRENT": needinfo,
         "VERSION_2": needinfo,
   },
    # Specified by the first time on the current docs
    "NEED_ID_C": {
         "CURRENT": needinfo,
   },
   # Deleted on the current docs 
    "NEED_ID_D": {
         "VERSION_2": needinfo,
   },
}
```
With data like this we could do a lot of awesome voodoo in future.

**Updated Import mechanism**
Currently `needimport` loads the data of a needs.json directly to the rst and makes all imported needs part of the current document.
But we would need a mechanism to load need.json data without adding this information directly to the docs. May be some option for `conf.py` or maybe even the `:hide:` option of `needimport` could be enough

**Load single need directive**
To show a specific need in a rst, we would need a directive like:

```rst
.. needload: NEED_ID : NEED_VERSION
```
or to show all needs based on a filter something like `needimport`, which data is not based on a given needs.json but is based on the complete, internal needlist.

Or do you think such a solution would be to over-engineered?

**Comment by tlovett** [comment]
What I was thinking was a little more modest I think.  First, the changelog would only be available to users of `needs_file`.  If you attempted to use it without that configuration, then the build would fail.

1.  Introduce a new directive:

```reStructuredText
.. needlog::
   :base-version: "0.2.0"
``

There might be other settings, but that would be the basics.  The directive would do a comparison between the base version in `needs_file` (if base-version isn't provided it could use the most recent version) and `needs_all_needs`.   It would only compare across a few key fields (this might be part of the configuration ultimately), and determine if a requirement has been added, removed, or changed.

If you're doing generated IDs this gets a little tricky because changing the hashed content will create a new ID, but I think people could live with that.

Similarly, the `need_info` object could get this comparison information as well although base version might have to move out to conf.py for that to work.  It could have the following items available.

* `change_type`  - added, removed, updated (NOTE: I need to think about removed a bit more because inserting the old version in to need_all_needs would probably mess up some default behavior)
* `base` - This would be present if updated, and give you access to the other fields from the old version of the requirement.
```

So you cold create a table like this:

```
.. needtable::
  :columns: id,change_type,content, base.content
```

I do agree about the ability to do things like diffs would be powerful, and injecting into the diffs themselves would be useful in some cases, but I think a modest first step towards that is available just be bringing forward the import of the `needs_file` in the build lifecycle and performing some comparisons once the doctree is ready.   

Thoughts?

Another more general purpose solution I'm looking at is being able to create Jinja2 templates in our RST documents in a directive.   I haven't found a community version that does what I want though. 

1. It should have access to the full environment - not a specific dict of contexts for the template rendering like sphinx-jinja.  This would enable me to pull in needs_all_needs into the template.

2. I also want the templates themselves to be wrapped in a pending element and delay processing (much like `:contents:`.  The transform would then render the template, and make use of data that collected during the build process

3. It needs some ability to have macros and load templates externally. 

I don't see that being part of needs per se, but it's something I think I could use to solve this use case as well.

**Comment by danwos** [comment]
You are right, my ideas are too complex/big for this case. Let's keep it more modest.

So maybe I misunderstood you, but I'm personal not a big fan of directives like `.. needlog::`, if they do not have any textual representation in the final document.
In this case I would prefer something inside `conf.py`.

But I think a nice solution is nearly available, if we have **dynamic option** and **global options** from #57.
In this case, we would only need 2-3 things:
**A global option in conf.py**
```python
needs_global_options = [{
  'need_filter' = "id is True"
  'change_type' : [[log_changed(base_version="0.2.0")]],
  'change_base': [[log_base(base_version="0.2.0")],
  },
]
```
**A dynamic var function for comparison**

```python
def log_changed(need, all_needs, *args, **kwargs):
    if "base_version" in kwargs:
       # load needs from needs.json
       # Compare needs
       # calc return value
    
   return change_type 

def log_base(need, all_needs, *args, **kwargs):
    ...
```

For sure the function should somehow save the imported and compared needs, so that this is done only once.
And `log_changed()` could also set the option of `change_base()` for the given need.
But I like the way that all manipulations on need options are documented in `needs_global_options`.
So as developer/user you always know what gets changed by taking a look into your `conf.py` and do not need to know the internal details of a used **dynamic var function**.

If #57 is available, then a lot of upcoming features will be implemented by using this mechanism.
So there will be some kind of a sphinx-needs collection for **dynamic var functions** and so `log_changed` could be the first function of this collection.

For sure all names are just rough ideas, feel free to make own suggestions.

What do you think?
In my understanding this should be a very clean and easy maintainable solution.
Or have I missed some points?

**Comment by danwos** [comment]
Ohh and for sure your Jinja proposal would work.
But in my experience working with Jinja templates to create RST files/content can be really hard, because RST is often based on a clean structure via whitespaces and Jinja has some functions to work with whitespaces, but in some scenarios they often delete too much of them or let them untouched.
For instance the sphinx-needs template: I tried a lot to have all whitespaces and lines deleted, if e.g. a status is not given. But there are still some scenarios where a blank line is left in rst and therefore also in the generated content...

**Comment by tlovett** [comment]
I agree that #57 would be a beneficial starting point for this feature.  I guess in your example, then the user would just leverage the the existing directives such as `needtable` to render the change log.  Is that what you were thinking?  

As for the change type and change base, are you proposing that this feature be closed and the user would just roll their own using #57 or would there be some built-in support for this?

**Comment by danwos** [comment]
> then the user would just leverage the the existing directives such as needtable to render the change log. Is that what you were thinking?

Yes, except the case some kind of a special layout of presentation is needed.
But if this is the case, we should create a new issue.

> are you proposing that this feature be closed and the user would just roll their own using #57 or would there be some built-in support for this?

No no, this should be realized as built-in function. So no need for the users to define their own functions for this case.
Just use #57/#58 and call the build-in function.
So just configure it:
```python
needs_global_options = {
  'change_type' : log_changed(base_version="0.2.0"),
  'change_base': log_base(base_version="0.2.0",
  }
```
and e.g. `log_changed` is already available.

For sure we need to discuss what kind of arguments are needed, e.g. needs_json_path, base_version, .... 
So any ideas, how the perfect function would look for you and your case?

**Comment by tlovett** [comment]
I'm good with setting aside a custom display for the change log as part of this feature.  In my mind, I was thinking of the configuration of this would be a little more direct vs. the user needing to use `needs_global_options`.

Instead in `conf.py` you'd configure it this way.

```Python

needs_file = 'path/to/needs.json'
needs_file_base_version = '0.2.0'
```

If you specify both of those, then the `change_type` would be available in columns and filters.  If `change_type` is `changed` (or `modified` we can decide on the right names), then the `previous_need` attribute would have the prior `need_info`.  

My rationale for doing it this way vs. having the user manually configure it in `needs_global_options` is 

1. I think this is a little easier for the users, but doesn't preclude them doing something more custom via global options
2. We'd control the standard metadata and its format if at some point we did want to layer on some of the features discussed earlier in this issue (change log directive, dynamic diffs, etc.)

Thoughts?

**Comment by danwos** [comment]
Puhh, both options have some good/bad arguments.

For me setting it via `global-options` has more advantages:

* all additional need-options are registered and viewable in `global options`.
  No black magic in the background, which adds additional options -> more Transparency
* option names are freely selectable. E.g. needed for different languages
* Support on single needs, because this can be used directly in rst files.
  -> May be an upcoming use case in my projects.
* Support for multiple needs.json files.
  -> E.g. we are importing multiple json-files from libraries-docs to get a common view on a master document.
* Support for multiple versions
  -> For some needs the comparison must be made with `1.0`, for others with `0.9`.
* Support for multiple and parallel version comparison.
   --> You could have `change_type_v1` and `change_type_v1.2` on a single need.

I agree, that using `global options` + a given `dynamic value` is a little bit too complex for a single function/use case.
But I'm pretty sure we will use and provide `dynamic values` for a bunch of other upcoming feature requests in the future. And therefore this kind of configuration will become somehow "common/familiar" for this kind of extensions.

So I would prefer to **not** provide additional `conf.py` parameters and go with above proposal.
Counter-arguments?

**Comment by tlovett** [comment]
I don't have any new counter-arguments really beyond what I previously shared.  Ultimately it's a trade-off between flexibility and end-user convenience.   This is workable from my perspective.  If we introduce additional features that can make use of this change tracking we'll just have to be explicit on the names and conventions those expect or make them configurable.

**Comment by danwos** [comment]
Okay, I feel more comfortable  with a realisation via #57 und #58.
Let's go that way and I will check how the feedback in my and others projects is.

**Comment by danwos** [comment]
Done / doable via #57 and #58.
I close this issue.
